doc/nat/nat_design_spec.md

```diff
@@ -0,0 +1,1550 @@
+
+### 3.3.3.5 Max limit on the NAT entries
+NatOrch keeps track of the total number of static + dynamic entries created and limits the NAT'ted conntrack entries in the kernel to the supported max limit in the hardware. When a new translation notification is received beyond the maximum limit (due to new traffic that is being NAT'ted), the conntrack entry is removed in the kernel.
+ 
+### 3.3.3.5 Block diagram
+The high level block diagram for NAT in SONiC is captured below.
+ 
+![NAT block diagram](images/Nat_block_diagram.png)
+ 
+## 3.4 Linux Integration
+### 3.4.1 IPtables
+In SONiC, the core NAT logic which involves the (IP + port) translation dynamically is the iptables module in the kernel.
+The iptables uses the Netfilter hooks in the kernel to execute its NAT'ting rules at different points in the packet forwarding path.
+
+The important hooks that iptables registers with Netfilter are the ones for PREROUTING and POSTROUTING. The DNAT rules if available are applied in the PREROUTING chain and SNAT rules if available are applied in the POSTROUTING chain.
+
+The packets that are going from one zone to another zone in the hardware are trapped to CPU as NAT miss packets if there are no matching NAT entries in the hardware. Such packets are received by the KNET kernel module and passed on to the netdev interface (corresponding to the interface the packet came on) in the Network stack.
+
+Before the L3 forwarding is done in the Kernel, the iptables rules in PREROUTING chain are applied. If there are any DNAT rules that the packets match against, they translate the DIP and L4 port to the IP and L4 port as applicable. In this process, the DNAT state and DIP of the connection entry in the connection tracking table are updated.
+
+After the L3 forwarding is done and before the packet is about to be sent out, the rules in the POSTROUTING chain are applied which do the SNAT (translate the SIP + L4 port to the public IP + L4 port). In this process, the SNAT state and SIP of the connection entry in the connection tracking table are updated.
+
+As long as the hardware entry is not installed, the NAT translation and forwarding is done in the Linux kernel using the iptables and conntrack entries.
+
+#### 3.4.1.1 No NAT Zones in the Kernel
+Similar to the zone per L3 interface that is programmed in the hardware, there is no concept of NAT zone on the interface in the Kernel. As a result, no zone checks can be done in the kernel when doing NAT. Matching is done against the outbound interface for the POSTROUTING/SNAT traffic and matching against the inbound interface for the PREROUTING/DNAT traffic. The outgoing interface parameter should be passed to the iptables rules to ensure that the NAT is done for the software routed traffic in the kernel only when the packet is sent on the interface corresponding to the outside zone. Without the interface parameter, any software routed packet can get NAT'ted.
+
```

Comment:
Would be good to improve the clarify of this sentence " The outgoing interface parameter should be passed to the iptables rules to ensure that the NAT is done for the software routed traffic in the kernel only when the packet is sent on the interface corresponding to the outside zone."  - what parameter is being passed.  Here it seems you're talking about packets received from downlinks belonging to one zone, to an uplink belonging to a different zone-id.  What's being checked to ensure "packet is sent on the interface corresponding to the outside zone"?

Reply:
Current design is as below:
For normal interfaces (non loopback interfaces) that are used as Public IP for NAT, interface parameter is passed when programming the iptables so that if the packets are going out on that interface, iptables apply the dynamic nat pool.
for eg., iptables -t nat -A POSTROUTING -j SNAT ... -o <Ethernetxx> --to-source <ip+port pool>

For loopback interface whose IP is used as NAT public IP, the packets can go on any interface that are in the same zone as the loopback interface. So in this case, interface parameter is not passed to the iptables, and hence it is handled differently. The iptables mangles table rules set the MARK value to the zone value when a packet is going out on an interface. And in the iptables nat table rules for loopback, we match on the MARK value that is matching the loopback zone value.

Sure, will update this section for clarity.

If we want the same behavior on normal interface too (to not just match only on the outgoing interface) as in the case of the loopback interface, we will be restricted to configuring only 1 pool for 1 outside zone.